src/Parallel/Actions/SetupDataBox.hpp

```diff
@@ -0,0 +1,132 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <tuple>
+#include <type_traits>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Actions {
+/// \cond
+struct SetupDataBox;
+/// \endcond
+
+namespace detail {
+CREATE_HAS_TYPE_ALIAS(simple_tags)
+CREATE_HAS_TYPE_ALIAS_V(simple_tags)
+CREATE_HAS_TYPE_ALIAS(compute_tags)
+CREATE_HAS_TYPE_ALIAS_V(compute_tags)
+
+template <typename Action, typename enable = std::bool_constant<true>>
+struct optional_simple_tags;
+
+template <typename Action>
+struct optional_simple_tags<
+    Action, std::bool_constant<detail::has_simple_tags_v<Action>>> {
+  using type = typename Action::simple_tags;
+};
+
+template <typename Action>
+struct optional_simple_tags<
+    Action, std::bool_constant<not detail::has_simple_tags_v<Action>>> {
+  using type = tmpl::list<>;
+};
+
+template <typename Action, typename enable = std::bool_constant<true>>
+struct optional_compute_tags;
+
+template <typename Action>
+struct optional_compute_tags<
+    Action, std::bool_constant<detail::has_compute_tags_v<Action>>> {
+  using type = typename Action::compute_tags;
+};
+
+template <typename Action>
+struct optional_compute_tags<
+    Action, std::bool_constant<not detail::has_compute_tags_v<Action>>> {
+  using type = tmpl::list<>;
+};
+
+template <typename ActionList>
+struct get_action_list_simple_tags;
+
+template <typename... ActionList>
+struct get_action_list_simple_tags<tmpl::list<ActionList...>> {
+  using type = tmpl::remove_duplicates<tmpl::flatten<
+      tmpl::list<typename optional_simple_tags<ActionList>::type...>>>;
```

Comment:
I think this is just `tmpl::transform<ActionList, tmpl::bind<optional_simple_tags, tmpl::_1>>`. Then you can make `get_action_list_simple_tags` a type alias instead of a struct.